keps/sig-cluster-lifecycle/kubeadm/4656-add-kubelet-instance-configuration/README.md

```diff
@@ -0,0 +1,287 @@
+
+**GA(v1.34)**:
+
+* `kubeadm upgrade apply/node` will use `/var/lib/kubelet/instace-config.yaml` and remove CRI annotations from Node object and remove `--container-runtime-endpoint` args in `/var/lib/kubelet/kubeadm-flags.env` file
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+At least the following kubeadm packages will require updates and new unit tests:
+
+* `cmd/kubeadm/app/cmd/phases/init`
+
+* `cmd/kubeadm/app/phases/kubelet`
+
+* `cmd/kubeadm/app/phases/upgrade`
+
+* `cmd/kubeadm/app/cmd/phases/join`
+
+##### Integration tests
+
+- N/A
+
+##### e2e tests
+
+* A new e2e test will be added by using the kinder tool.
+
+### Graduation Criteria
+
+#### Alpha
+
+- The upgrade will still respect the CRI annotation.
+- Use `NodeLocalCRISocket` feature gate to implement features.
+- Add corresponding e2e tests.
+
+#### Beta
+
+* Make feature gate to be enabled by default.
+
+- Use instance kubelet configuration, ignore CRI annotations.
+- Gather feedback from developers and surveys.
+
+#### GA
+
+- Remove the CRI annotation during upgrade(this may be not urgent or have to).
+
+### Upgrade / Downgrade Strategy
+
+**Alpha**: Users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to enable the `NodeLocalCRISocket` feature gate before calling kubeadm upgrade apply, which will allow a `/var/lib/kubelet/instace-config.yaml` to be generated and overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` with it.
+
+**Beta**: Users can modify `ClusterConfiguration` to disable the feature gate during upgrades. This will allow them to continue using the CRI socket annotation on nodes.
+
+**GA**: Users can no longer patch ClusterConfiguration to opt out of the feature and it will be locked to be enabled by default.
+
+### Version Skew Strategy
+
+kubeadm will continue to skew from kubelet for three versions. The `ContainerRuntimeEndpoint` field in `KubeletConfiguration` was [introduced](https://github.com/kubernetes/kubernetes/pull/112136) in v1.27, so when we overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` through `/var/lib/kubelet/instace-config.yaml`, it will be supported on all versions of the kubelet within the skew.
+
+## Implementation History
+
+- 2024-05-23: Initial draft KEP
+
+## Drawbacks
+
+* This KEP will bring a breaking change. Many potential users (I believe it is the majority) still rely on the `kubeadm.alpha.kubernetes.io/cri-socket` annotation or the `/var/lib/kubelet/kubeadm-flags.env` file to declare the CRI socket endpoint on the Node, because many users are familiar with them.
```

Comment:
most users don't really "rely" on it as it's an implementation detail.
some users do read / write it an "rely" on it, yes.